src/Middleware.php

```diff
@@ -186,7 +186,24 @@ public static function retry(
             return new RetryMiddleware($decider, $delay, $handler, $stats);
         };
     }
-
+    /**
+    * Middleware wrapper function that adds an invocation id header to requests.
+    * This is a uniquely generated UUID to identify initial and subsequent retries as part of a complete request lifecycle.
+    *
+    * @return callable
+    */
+    public static function invocationID()
+    {
+        return function (callable $handler) {
+            return function (
+                CommandInterface $command,
+                RequestInterface $request = null
+            ) use ($handler) {
+                $request = $request->withHeader('aws-sdk-invocation-id', uniqid(gethostname(), true));
```

Comment:
The host name should be normalized to a fixed length and character set. Maybe you could use the md5 or crc32 of the hostname instead?